### PR TITLE
TW-12305 - logger: Gör en version av php-loggern som är PHP 7.4-kompatibel

### PR DESCRIPTION
**Pull request by martinsiden**
## Summary by CodeRabbit

* **Chores**
  * Added official support for PHP 7.4 alongside PHP 8.x; composer.lock is now ignored.

* **Refactor**
  * Reworked logging internals, log format/level representations, and span handling for clearer behavior and consistency.

* **Bug Fixes**
  * Unknown log severity values now raise an explicit error instead of failing silently; OTLP mapping corrected.

* **Tests**
  * Expanded PHPUnit coverage for logging, formats, levels, OTLP mapping, singleton behavior, and span timestamp precision.

### CODE REVIEW

**Comment by lilleman-tw** [comment]
Claude gnällde:

Blocker

1. composer.lock is incompatible with PHP 7.4. It pins doctrine/instantiator v2.1.0, which requires PHP ^8.4. Anyone following the README's composer install flow on 7.4 will fail. CI passes only because the workflow uses composer update, which silently re-resolves. Fix by re-locking against a PHP 7.4 platform constraint, pinning transitive dev deps in composer.json, or (cleanest for a library) dropping composer.lock from the repo.

Major

2. Enum factories are no longer singletons. LogLevel::error() === LogLevel::error() is now false. Native enums guaranteed identity; the backport returns a fresh instance every call. External callers using === (idiomatic) will silently break. Fix with cached singletons via ??=.
3. $name/$value are publicly mutable — readonly was dropped. Anyone can corrupt a LogLevel instance. Combined with #2, mutating one instance doesn't even propagate. Fix by making properties private with getters.
4. Silent default in OTLP severity mapping (CustomLogger.php:180-182). The original match threw UnhandledMatchError for unknown levels; the new switch falls through to severityNumber = 0 ("UNSPECIFIED"). Bad data now reaches Grafana Cloud instead of failing fast. Fix by throwing LogicException in the default branch.

Minor / Nit

5. CI matrix covers 7.4 + 8.4, but no test exercises the OTLP HTTP path on 7.4 — riskiest 7.4-specific code (cURL, JSON edge cases) is unverified.
6. API surface inconsistency: LogFormat has tryFrom(), LogLevel doesn't.
7. Pre-existing fd leak: fopen('php://stdout', 'w') per log call without fclose. Not introduced here, but worth noting (fwrite(STDOUT, ...) is the fix).
8. Pre-existing precision bug in Span.php:53-54,91: (int)microtime(true) * 1e9 truncates sub-second precision before multiplying. Trace timestamps are off by up to ~1s Fix: cast after multiplying.
9. Test testLogMethodAcceptsLogLevelInstanceDirectly doesn't pin the new singleton/non-singleton contract — worth adding once #2 is decided.

File: tests/LogFormatTest.php [hunk 1]
```diff
@@ -31,4 +31,25 @@ public function testTryFromUnknownValueReturnsNull(): void
...
+    }
+
+    public function testInstancesRejectMutation(): void
+    {
+        // Singletons must not be mutable, otherwise a write through one
+        // reference would corrupt every cached caller.
+        $format = LogFormat::text();
+
+        $this->expectException(\LogicException::class);
+        $format->value = 'mutated';
```
[the remark this answers is not in the document]
Comment (martinsiden) [reply tests/LogFormatTest.php:57]: @coderabbitai PHPStan isn't currently part of this project's toolchain; suppressions would be premature. Will revisit if/when a static analyzer is added in a follow-up.

File: tests/LogLevelTest.php [hunk 2]
```diff
@@ -7,24 +7,85 @@
...
+
+    public function testInstancesRejectMutation(): void
+    {
+        // Native enum cases are readonly; the polyfill must prevent writes too,
+        // otherwise mutating one call site would corrupt the cached singleton
+        // for every other caller.
+        $level = LogLevel::error();
+
+        $this->expectException(\LogicException::class);
+        $level->name = 'mutated';
```
[the remark this answers is not in the document]
Comment (martinsiden) [reply tests/LogLevelTest.php:68]: @coderabbitai PHPStan isn't currently part of this project's toolchain; suppressions would be premature. Will revisit if/when a static analyzer is added in a follow-up.

**Review by Eismar — APPROVED** [review]

**Review by lilleman-tw — APPROVED** [review]
